zerodep/src/main/java/com/lightbend/microprofile/reactive/streams/zerodep/StageInlet.java

```diff
@@ -90,23 +90,30 @@ public void onUpstreamFailure(Throwable error) {
 
   /**
    * Indicates that an element has been pushed. The element can be received using {@link StageInlet#grab()}.
+   * <p>
+   * If this throws an exception, the error will be passed to {@link #onUpstreamFailure(Throwable)}, upstream will
+   * be cancelled, and the stage will not receive any further signals.
    */
   void onPush();
 
   /**
-   * Indicates that upstream has completed the stream. No signals may be sent to the inlet after this has been invoked.
-   *
-   * This must be very careful not to throw an exception. If it does, then the signal to complete will not reach
-   * downstream.
+   * Indicates that upstream has completed the stream. Unless this throws an exception, no signals may be sent to the
+   * inlet after this has been invoked.
+   * <p>
+   * If this throws an exception, the error will be passed to {@link #onUpstreamFailure(Throwable)}, upstream will
+   * be cancelled, and the stage will not receive any further signals. Stages should be careful to ensure that if they
+   * do throw from this method, that they are ready to receive that exception from {@code onUpstreamFailure}.
    */
   void onUpstreamFinish();
 
   /**
-   * Indicates that upstream has completed the stream with a failure. No signals may be sent to the inlet after this has
-   * been invoked.
-   *
-   * This must be very careful not to throw an exception. If it does, then the signal to complete will not reach
-   * downstream.
+   * Indicates that upstream has completed the stream with a failure. Once this has been invoked, no other signals
+   * will be sent to this listener.
+   * <p>
+   * If this throws an exception, the entire stream will be shut down, since there's no other way to guarantee that
```

Comment:
> the entire stream will be shut down

No, why? Why is there an alternative "entire stream shut down" other than error and cancellation propagation?

Reply:
So let's say I have three stages, `1`, `2` and `3`, with connector `A` between `1` and `2`, and connector `B` between `2` and `3`. My graph looks like this:

```
A -> 1 -> B -> 2 -> C
```

Now let's say `A` fails, sending the failure to `1`. `1` passes the failure signal to `B` via `onUpstreamFailure`, but that in turn throws an exception. At this point, what does `1` do? There's no way to know if `B` passed the signal on to `2` before it threw the signal, indeed, `1` knows nothing of `2`'s existence, or of what if anything `B` should have done with the signal it sent it. If it didn't, the stream will now never terminate, and we'll potentially have a resource leak since stage `C` will never get shut down. So, as a safety, if `onUpstreamFailure` throws, we trigger the non graceful entire stream termination, where every in/out port that isn't already finished is terminated. This will ensure that `C` receives the `onUpstreamFailure` signal.

Reply:
To clarify: The "no way to know" things all only explain this specific implementation. There are ways to know, and e.g. Akka Streams would get that thrown exception and fail C with it. I see that if you're trying to go for a minimal not that user friendly implementation it could be ok I guess. At least it shows where Akka Streams provide nicer error propagation

Reply:
And it should be stressed, this should never happen during normal operation, if `B` throws, that's a bug. This is just being defensive, ensuring that bugs don't cause resource leaks.

Reply:
Yes, it is specific to this implementation - but note that all of these interfaces are 100% implementation specific, they're package private, no contract or guarantees, if the implementation changes the interface may too and the documentation will as well. The javadocs only exists to explain to someone trying to understand the implementation and how it works. Keeping some understanding of the structure of the graph outside of the wired graph itself would greatly increase the complexity I think, and I don't think gives much value, since the error case indicates a bug - even Akka streams can't guarantee that exception caused by bugs in itself will always flow downstream, in some cases, the actor will crash, which is essentially what is being specified here.